src/transaction.rs

```diff
@@ -48,10 +49,40 @@ pub struct Transaction {
     pub crossover: Option<Crossover>,
     /// Serialized proof of the `Execute` circuit for this transaction.
     pub proof: Vec<u8>,
+    /// An optional execution carried by the transaction.
+    pub exec: Option<Execution>,
+}
+
+/// An execution is either a call to a contract, or a contract deployment.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "rkyv-impl",
+    derive(Archive, Serialize, Deserialize),
+    archive_attr(derive(bytecheck::CheckBytes))
+)]
+pub enum Execution {
     /// A call to a contract. The `Vec<u8>` must be an `rkyv`ed representation
     /// of the data the contract expects, and the `String` the name of the
-    /// function to call.
-    pub call: Option<(ModuleId, String, Vec<u8>)>,
+    /// contract function to call.
+    Call {
+        /// The contract to be called
+        contract: ModuleId,
+        /// The function to call in the contract
+        fn_name: String,
+        /// The argument to the function to call
+        fn_arg: Vec<u8>,
+    },
+    /// A contract deployment.
+    Deploy {
+        /// The owner of the contract
+        owner: Box<PublicKey>, /* boxed due to a large size difference
+                                * between variants */
```

Comment:
I'm not sure what you mean with the large size difference between variants. To me it seems that the `PublicKey`, with it's fixed 64 bytes is not that much bigger than the 32 bytes of the `ModuleId` and the rest of the fields are pointers. Am I missing something?

Reply:
A `PublicKey` is 320 bytes in memory - there's two extended points in there. So between that and 32, there quite some bytes.

Reply:
oh wait, I was thinking wrong, you are right!